File: app/src/models/popup.ts
import {
  Repository,
  RepositoryWithGitHubRepository,
  RepositoryWithForkedGitHubRepository,
} from './repository'
import { PullRequest } from './pull-request'
import { Branch } from './branch'
import { ReleaseNote, ReleaseSummary } from './release-notes'
import { IRemote } from './remote'
import { RetryAction } from './retry-actions'
import { WorkingDirectoryFileChange } from './status'
import { PreferencesTab } from './preferences'
import { Commit, CommitOneLine, ICommitContext } from './commit'
import { IStashEntry } from './stash-entry'
import { Account } from '../models/account'
import { Progress } from './progress'
import { ITextDiff, DiffSelection, ImageDiffType } from './diff'
import { RepositorySettingsTab } from '../ui/repository-settings/repository-settings'
import { ICommitMessage } from './commit-message'
import { Author, UnknownAuthor } from './author'
import { IRefCheck } from '../lib/ci-checks/ci-checks'
import { GitHubRepository } from './github-repository'
import { ValidNotificationPullRequestReview } from '../lib/valid-notification-pull-request-review'
import { UnreachableCommitsTab } from '../ui/history/unreachable-commits-dialog'
import { IAPIComment } from '../lib/api'

export enum PopupType {
  RenameBranch = 'RenameBranch',
  DeleteBranch = 'DeleteBranch',
  DeleteRemoteBranch = 'DeleteRemoteBranch',
  ConfirmDiscardChanges = 'ConfirmDiscardChanges',
  Preferences = 'Preferences',
  RepositorySettings = 'RepositorySettings',
  AddRepository = 'AddRepository',
  CreateRepository = 'CreateRepository',
  CloneRepository = 'CloneRepository',
  CreateBranch = 'CreateBranch',
  SignIn = 'SignIn',
  About = 'About',
  InstallGit = 'InstallGit',
  PublishRepository = 'PublishRepository',
  Acknowledgements = 'Acknowledgements',
  UntrustedCertificate = 'UntrustedCertificate',
  RemoveRepository = 'RemoveRepository',
  TermsAndConditions = 'TermsAndConditions',
  PushBranchCommits = 'PushBranchCommits',
  CLIInstalled = 'CLIInstalled',
  GenericGitAuthentication = 'GenericGitAuthentication',
  ExternalEditorFailed = 'ExternalEditorFailed',
  OpenShellFailed = 'OpenShellFailed',
  InitializeLFS = 'InitializeLFS',
  LFSAttributeMismatch = 'LFSAttributeMismatch',
  UpstreamAlreadyExists = 'UpstreamAlreadyExists',
  ReleaseNotes = 'ReleaseNotes',
  DeletePullRequest = 'DeletePullRequest',
  OversizedFiles = 'OversizedFiles',
  CommitConflictsWarning = 'CommitConflictsWarning',
  PushNeedsPull = 'PushNeedsPull',
  ConfirmForcePush = 'ConfirmForcePush',
  StashAndSwitchBranch = 'StashAndSwitchBranch',
  ConfirmOverwriteStash = 'ConfirmOverwriteStash',
  ConfirmDiscardStash = 'ConfirmDiscardStash',
  ConfirmCheckoutCommit = 'ConfirmCheckoutCommit',
  CreateTutorialRepository = 'CreateTutorialRepository',
  ConfirmExitTutorial = 'ConfirmExitTutorial',
  PushRejectedDueToMissingWorkflowScope = 'PushRejectedDueToMissingWorkflowScope',
  SAMLReauthRequired = 'SAMLReauthRequired',
  CreateFork = 'CreateFork',
  CreateTag = 'CreateTag',
  DeleteTag = 'DeleteTag',
  LocalChangesOverwritten = 'LocalChangesOverwritten',
  ChooseForkSettings = 'ChooseForkSettings',
  ConfirmDiscardSelection = 'ConfirmDiscardSelection',
  MoveToApplicationsFolder = 'MoveToApplicationsFolder',
  ChangeRepositoryAlias = 'ChangeRepositoryAlias',
  ThankYou = 'ThankYou',
  CommitMessage = 'CommitMessage',
  MultiCommitOperation = 'MultiCommitOperation',
  WarnLocalChangesBeforeUndo = 'WarnLocalChangesBeforeUndo',
  WarningBeforeReset = 'WarningBeforeReset',
  InvalidatedToken = 'InvalidatedToken',
  AddSSHHost = 'AddSSHHost',
  SSHKeyPassphrase = 'SSHKeyPassphrase',
  SSHUserPassword = 'SSHUserPassword',
  PullRequestChecksFailed = 'PullRequestChecksFailed',
  CICheckRunRerun = 'CICheckRunRerun',
  WarnForcePush = 'WarnForcePush',
  DiscardChangesRetry = 'DiscardChangesRetry',
  PullRequestReview = 'PullRequestReview',
  UnreachableCommits = 'UnreachableCommits',
  StartPullRequest = 'StartPullRequest',
  Error = 'Error',
  InstallingUpdate = 'InstallingUpdate',
  TestNotifications = 'TestNotifications',
  PullRequestComment = 'PullRequestComment',
  UnknownAuthors = 'UnknownAuthors',
  ConfirmRepoRulesBypass = 'ConfirmRepoRulesBypass',
  TestIcons = 'TestIcons',
}

interface IBasePopup {
  /**
   * Unique id of the popup that it receives upon adding to the stack.
   */
  readonly id?: string
}

export type PopupDetail =
  | { type: PopupType.RenameBranch; repository: Repository; branch: Branch }
  | {
      type: PopupType.DeleteBranch
      repository: Repository
      branch: Branch
      existsOnRemote: boolean
    }
  | {
      type: PopupType.DeleteRemoteBranch
      repository: Repository
      branch: Branch
    }
  | {
      type: PopupType.ConfirmDiscardChanges
      repository: Repository
      files: ReadonlyArray<WorkingDirectoryFileChange>
      showDiscardChangesSetting?: boolean
      discardingAllChanges?: boolean
    }
  | {
      type: PopupType.ConfirmDiscardSelection
      repository: Repository
      file: WorkingDirectoryFileChange
      diff: ITextDiff
      selection: DiffSelection
    }
  | { type: PopupType.Preferences; initialSelectedTab?: PreferencesTab }
  | {
      type: PopupType.RepositorySettings
      repository: Repository
      initialSelectedTab?: RepositorySettingsTab
    }
  | { type: PopupType.AddRepository; path?: string }
  | { type: PopupType.CreateRepository; path?: string }
  | {
      type: PopupType.CloneRepository
      initialURL: string | null
    }
  | {
      type: PopupType.CreateBranch
      repository: Repository
      initialName?: string
      targetCommit?: CommitOneLine
    }
  | {
      type: PopupType.SignIn
      isCredentialHelperSignIn?: boolean
      credentialHelperUrl?: string
    }
  | { type: PopupType.About }
  | { type: PopupType.InstallGit; path: string }
  | { type: PopupType.PublishRepository; repository: Repository }
  | { type: PopupType.Acknowledgements }
  | {
      type: PopupType.UntrustedCertificate
      certificate: Electron.Certificate
      url: string
    }
  | { type: PopupType.RemoveRepository; repository: Repository }
  | { type: PopupType.TermsAndConditions }
  | {
      type: PopupType.PushBranchCommits
      repository: Repository
      branch: Branch
      unPushedCommits?: number
    }
  | { type: PopupType.CLIInstalled }
  | {
      type: PopupType.GenericGitAuthentication
      remoteUrl: string
      username?: string
      onSubmit: (username: string, password: string) => void
      onDismiss: () => void
    }
  | {
      type: PopupType.ExternalEditorFailed
      message: string
      suggestDefaultEditor?: boolean
      openPreferences?: boolean
    }
  | { type: PopupType.OpenShellFailed; message: string }
  | { type: PopupType.InitializeLFS; repositories: ReadonlyArray<Repository> }
  | { type: PopupType.LFSAttributeMismatch }
  | {
      type: PopupType.UpstreamAlreadyExists
      repository: Repository
      existingRemote: IRemote
    }
  | {
      type: PopupType.ReleaseNotes
      newReleases: ReadonlyArray<ReleaseSummary>
    }
  | {
      type: PopupType.DeletePullRequest
      repository: Repository
      branch: Branch
      pullRequest: PullRequest
    }
  | {
      type: PopupType.OversizedFiles
      oversizedFiles: ReadonlyArray<string>
      context: ICommitContext
      repository: Repository
    }
  | {
      type: PopupType.CommitConflictsWarning
      /** files that were selected for committing that are also conflicted */
      files: ReadonlyArray<WorkingDirectoryFileChange>
      /** repository user is committing in */
      repository: Repository
      /** information for completing the commit */
      context: ICommitContext
    }
  | {
      type: PopupType.PushNeedsPull
      repository: Repository
    }
  | {
      type: PopupType.ConfirmForcePush
      repository: Repository
      upstreamBranch: string
    }
  | {
      type: PopupType.StashAndSwitchBranch
      repository: Repository
      branchToCheckout: Branch
    }
  | {
      type: PopupType.ConfirmOverwriteStash
      repository: Repository
      branchToCheckout: Branch | null
    }
  | {
      type: PopupType.ConfirmDiscardStash
      repository: Repository
      stash: IStashEntry
    }
  | {
      type: PopupType.ConfirmCheckoutCommit
      repository: Repository
      commit: CommitOneLine
    }
  | {
      type: PopupType.CreateTutorialRepository
      account: Account
      progress?: Progress
    }
  | {
      type: PopupType.ConfirmExitTutorial
    }
  | {
      type: PopupType.PushRejectedDueToMissingWorkflowScope
      rejectedPath: string
      repository: RepositoryWithGitHubRepository
    }
  | {
      type: PopupType.SAMLReauthRequired
      organizationName: string
      endpoint: string
      retryAction?: RetryAction
    }
  | {
      type: PopupType.CreateFork
      repository: RepositoryWithGitHubRepository
      account: Account
    }
  | {
      type: PopupType.CreateTag
      repository: Repository
      targetCommitSha: string
      initialName?: string
      localTags: Map<string, string> | null
    }
  | {
      type: PopupType.DeleteTag
      repository: Repository
      tagName: string
    }
  | {
      type: PopupType.ChooseForkSettings
      repository: RepositoryWithForkedGitHubRepository
    }
  | {
      type: PopupType.LocalChangesOverwritten
      repository: Repository
      retryAction: RetryAction
      files: ReadonlyArray<string>
    }
  | { type: PopupType.MoveToApplicationsFolder }
  | { type: PopupType.ChangeRepositoryAlias; repository: Repository }
  | {
      type: PopupType.ThankYou
      userContributions: ReadonlyArray<ReleaseNote>
      friendlyName: string
      latestVersion: string | null
    }
  | {
      type: PopupType.CommitMessage
      coAuthors: ReadonlyArray<Author>
      showCoAuthoredBy: boolean
      commitMessage: ICommitMessage | null
      dialogTitle: string
      dialogButtonText: string
      prepopulateCommitSummary: boolean
      repository: Repository
      onSubmitCommitMessage: (context: ICommitContext) => Promise<boolean>
    }
  | {
      type: PopupType.MultiCommitOperation
      repository: Repository
    }
  | {
      type: PopupType.WarnLocalChangesBeforeUndo
      repository: Repository
      commit: Commit
      isWorkingDirectoryClean: boolean
    }
  | {
      type: PopupType.WarningBeforeReset
      repository: Repository
      commit: Commit
    }
  | {
      type: PopupType.InvalidatedToken
      account: Account
    }
  | {
      type: PopupType.AddSSHHost
      host: string
      ip: string
      keyType: string
      fingerprint: string
      onSubmit: (addHost: boolean) => void
    }
  | {
      type: PopupType.SSHKeyPassphrase
      keyPath: string
      onSubmit: (
        passphrase: string | undefined,
        storePassphrase: boolean
      ) => void
    }
  | {
      type: PopupType.SSHUserPassword
      username: string
      onSubmit: (password: string | undefined, storePassword: boolean) => void
    }
  | {
      type: PopupType.PullRequestChecksFailed
      repository: RepositoryWithGitHubRepository
      pullRequest: PullRequest
      shouldChangeRepository: boolean
      checks: ReadonlyArray<IRefCheck>
    }
  | {
      type: PopupType.CICheckRunRerun
      checkRuns: ReadonlyArray<IRefCheck>
      repository: GitHubRepository
      prRef: string
      failedOnly: boolean
    }
  | { type: PopupType.WarnForcePush; operation: string; onBegin: () => void }
  | {
      type: PopupType.DiscardChangesRetry
      retryAction: RetryAction
    }
  | {
      type: PopupType.PullRequestReview
      repository: RepositoryWithGitHubRepository
      pullRequest: PullRequest
      review: ValidNotificationPullRequestReview
      shouldCheckoutBranch: boolean
      shouldChangeRepository: boolean
    }
  | {
      type: PopupType.UnreachableCommits
      selectedTab: UnreachableCommitsTab
    }
  | {
      type: PopupType.StartPullRequest
      prBaseBranches: ReadonlyArray<Branch>
      currentBranch: Branch
      defaultBranch: Branch | null
      externalEditorLabel?: string
      imageDiffType: ImageDiffType
      prRecentBaseBranches: ReadonlyArray<Branch>
      repository: Repository
      nonLocalCommitSHA: string | null
      showSideBySideDiff: boolean
      currentBranchHasPullRequest: boolean
    }
  | {
      type: PopupType.Error
      error: Error
    }
  | {
      type: PopupType.InstallingUpdate
    }
  | {
      type: PopupType.TestNotifications
      repository: RepositoryWithGitHubRepository
    }
  | {
      type: PopupType.PullRequestComment
      repository: RepositoryWithGitHubRepository
      pullRequest: PullRequest
      comment: IAPIComment
      shouldCheckoutBranch: boolean
      shouldChangeRepository: boolean
    }
  | {
      type: PopupType.UnknownAuthors
      authors: ReadonlyArray<UnknownAuthor>
      onCommit: () => void
    }
  | {
      type: PopupType.ConfirmRepoRulesBypass
      repository: GitHubRepository
      branch: string
      onConfirm: () => void
    }
  | {
      type: PopupType.TestIcons
    }

export type Popup = IBasePopup & PopupDetail

File: GitDesktop/GitDesktop/Models/Popup.swift
import Foundation

// MARK: - Popup (pruned)
// Port of `electron/app/src/models/popup.ts` with the AGENTS.md deletions
// applied: no GitHub (SignIn, PublishRepository, PRs, forks, SAML/re-auth,
// token, secret-scanning bypass API), no editor/terminal integration, no
// Copilot, no OS notifications. Closures in the original (`onSubmit`,
// `resolve`, …) become plain data here; Task 2's DialogHost wires them to
// `AppStore` actions.

public enum PopupType: String, Codable, Sendable {
    case renameBranch
    case deleteBranch
    case deleteRemoteBranch
    case confirmDiscardChanges
    case confirmDiscardSelection
    case preferences
    case repositorySettings
    case addRepository
    case createRepository
    case cloneRepository
    case createBranch
    case about
    case installGit
    case acknowledgements
    case shortcuts
    case untrustedCertificate
    case removeRepository
    case termsAndConditions
    case pushBranchCommits
    case cliInstalled
    case genericGitAuthentication
    case initializeLFS
    case lfsAttributeMismatch
    case upstreamAlreadyExists
    case releaseNotes
    case oversizedFiles
    case commitConflictsWarning
    case pushNeedsPull
    case confirmForcePush
    case stashAndSwitchBranch
    case confirmOverwriteStash
    case confirmDiscardStash
    case confirmCheckoutCommit
    case createTutorialRepository
    case confirmExitTutorial
    case createTag
    case deleteTag
    case localChangesOverwritten
    case moveToApplicationsFolder
    case changeRepositoryAlias
    case thankYou
    case commitMessage
    case multiCommitOperation
    case warnLocalChangesBeforeUndo
    case warningBeforeReset
    case addSSHHost
    case sshKeyPassphrase
    case sshUserPassword
    case warnForcePush
    case discardChangesRetry
    case unreachableCommits
    case error
    case installingUpdate
    case unknownAuthors
    case confirmCommitFilteredChanges
    case hookFailed
    case commitProgress
    case addWorktree
    case renameWorktree
    case deleteWorktree
    case deleteWorktreeFailed
}

public enum Popup: Sendable, Equatable, Identifiable {
    case renameBranch(repositoryID: Int, branchRef: String)
    case deleteBranch(repositoryID: Int, branchRef: String, existsOnRemote: Bool)
    case deleteRemoteBranch(repositoryID: Int, branchRef: String)
    case confirmDiscardChanges(repositoryID: Int, fileIDs: [String], showDiscardChangesSetting: Bool, discardingAllChanges: Bool)
    case confirmDiscardSelection(repositoryID: Int, fileID: String)
    case preferences(initialTab: String?)
    case repositorySettings(repositoryID: Int, initialTab: String?)
    case addRepository(path: String?)
    case createRepository(path: String?)
    case cloneRepository(initialURL: String?)
    case createBranch(repositoryID: Int, initialName: String?, targetCommitSHA: String?)
    case about
    case installGit(path: String)
    case acknowledgements
    case shortcuts
    case untrustedCertificate(host: String, certificateData: String)
    case removeRepository(repositoryID: Int)
    case termsAndConditions
    case pushBranchCommits(repositoryID: Int, branchRef: String, unPushedCommits: Int?)
    case cliInstalled
    case genericGitAuthentication(remoteURL: String, username: String?)
    case initializeLFS(repositoryIDs: [Int])
    case lfsAttributeMismatch
    case upstreamAlreadyExists(repositoryID: Int, existingRemoteName: String)
    case releaseNotes
    case oversizedFiles(repositoryID: Int, filePaths: [String])
    case commitConflictsWarning(repositoryID: Int, fileIDs: [String])
    case pushNeedsPull(repositoryID: Int)
    case confirmForcePush(repositoryID: Int, upstreamBranch: String)
    case stashAndSwitchBranch(repositoryID: Int, branchRef: String)
    case confirmOverwriteStash(repositoryID: Int, branchRef: String?)
    case confirmDiscardStash(repositoryID: Int, stashName: String)
    case confirmCheckoutCommit(repositoryID: Int, commitSHA: String)
    case createTutorialRepository
    case confirmExitTutorial
    case createTag(repositoryID: Int, targetCommitSHA: String, initialName: String?)
    case deleteTag(repositoryID: Int, tagName: String)
    case localChangesOverwritten(repositoryID: Int, files: [String])
    case moveToApplicationsFolder
    case changeRepositoryAlias(repositoryID: Int)
    case thankYou
    case commitMessage(repositoryID: Int, dialogTitle: String, dialogButtonText: String)
    case multiCommitOperation(repositoryID: Int)
    case warnLocalChangesBeforeUndo(repositoryID: Int, commitSHA: String, isWorkingDirectoryClean: Bool)
    case warningBeforeReset(repositoryID: Int, commitSHA: String)
    case addSSHHost(host: String, fingerprint: String)
    case sshKeyPassphrase(keyPath: String)
    case sshUserPassword(username: String)
    case warnForcePush(operation: String)
    case discardChangesRetry(repositoryID: Int)
    case unreachableCommits(repositoryID: Int)
    case error(message: String)
    case installingUpdate
    case unknownAuthors(repositoryID: Int, authors: [String])
    case confirmCommitFilteredChanges(fileCount: Int)
    case hookFailed(hookName: String, terminalOutput: String)
    case commitProgress
    case addWorktree(repositoryID: Int, initialBranchName: String?, initialWorktreeName: String?)
    case renameWorktree(repositoryID: Int, worktreePath: String)
    case deleteWorktree(repositoryID: Int, worktreePath: String)
    case deleteWorktreeFailed(repositoryID: Int, worktreePath: String, message: String)

    public var type: PopupType {
        switch self {
        case .renameBranch: return .renameBranch
        case .deleteBranch: return .deleteBranch
        case .deleteRemoteBranch: return .deleteRemoteBranch
        case .confirmDiscardChanges: return .confirmDiscardChanges
        case .confirmDiscardSelection: return .confirmDiscardSelection
        case .preferences: return .preferences
        case .repositorySettings: return .repositorySettings
        case .addRepository: return .addRepository
        case .createRepository: return .createRepository
        case .cloneRepository: return .cloneRepository
        case .createBranch: return .createBranch
        case .about: return .about
        case .installGit: return .installGit
        case .acknowledgements: return .acknowledgements
        case .shortcuts: return .shortcuts
        case .untrustedCertificate: return .untrustedCertificate
        case .removeRepository: return .removeRepository
        case .termsAndConditions: return .termsAndConditions
        case .pushBranchCommits: return .pushBranchCommits
        case .cliInstalled: return .cliInstalled
        case .genericGitAuthentication: return .genericGitAuthentication
        case .initializeLFS: return .initializeLFS
        case .lfsAttributeMismatch: return .lfsAttributeMismatch
        case .upstreamAlreadyExists: return .upstreamAlreadyExists
        case .releaseNotes: return .releaseNotes
        case .oversizedFiles: return .oversizedFiles
        case .commitConflictsWarning: return .commitConflictsWarning
        case .pushNeedsPull: return .pushNeedsPull
        case .confirmForcePush: return .confirmForcePush
        case .stashAndSwitchBranch: return .stashAndSwitchBranch
        case .confirmOverwriteStash: return .confirmOverwriteStash
        case .confirmDiscardStash: return .confirmDiscardStash
        case .confirmCheckoutCommit: return .confirmCheckoutCommit
        case .createTutorialRepository: return .createTutorialRepository
        case .confirmExitTutorial: return .confirmExitTutorial
        case .createTag: return .createTag
        case .deleteTag: return .deleteTag
        case .localChangesOverwritten: return .localChangesOverwritten
        case .moveToApplicationsFolder: return .moveToApplicationsFolder
        case .changeRepositoryAlias: return .changeRepositoryAlias
        case .thankYou: return .thankYou
        case .commitMessage: return .commitMessage
        case .multiCommitOperation: return .multiCommitOperation
        case .warnLocalChangesBeforeUndo: return .warnLocalChangesBeforeUndo
        case .warningBeforeReset: return .warningBeforeReset
        case .addSSHHost: return .addSSHHost
        case .sshKeyPassphrase: return .sshKeyPassphrase
        case .sshUserPassword: return .sshUserPassword
        case .warnForcePush: return .warnForcePush
        case .discardChangesRetry: return .discardChangesRetry
        case .unreachableCommits: return .unreachableCommits
        case .error: return .error
        case .installingUpdate: return .installingUpdate
        case .unknownAuthors: return .unknownAuthors
        case .confirmCommitFilteredChanges: return .confirmCommitFilteredChanges
        case .hookFailed: return .hookFailed
        case .commitProgress: return .commitProgress
        case .addWorktree: return .addWorktree
        case .renameWorktree: return .renameWorktree
        case .deleteWorktree: return .deleteWorktree
        case .deleteWorktreeFailed: return .deleteWorktreeFailed
        }
    }

    /// Stable identity for sheet presentation. Error popups stack (they are
    /// never deduped); all other types have one instance per type.
    public var id: String {
        switch self {
        case .error(let message): return "error-\(abs(message.hashValue))"
        default: return "popup-\(type.rawValue)"
        }
    }
}
